drivers/dsp0401/dsp0401.c

```diff
@@ -0,0 +1,251 @@
+    { 0x68, 0x68 },  /*  78:   N  */
+    { 0xAC, 0xAC },  /*  79:   0  */
+    { 0x1C, 0xB8 },  /*  80:   P  */
+    { 0xEC, 0xAC },  /*  81:   Q  */
+    { 0x5C, 0xB8 },  /*  82:   R  */
+    { 0xB4, 0xB4 },  /*  83:   S  */
+    { 0x05, 0x81 },  /*  84:   T  */
+    { 0xA8, 0x2C },  /*  85:   U  */
+    { 0x02, 0x2A },  /*  86:   V  */
+    { 0x68, 0x2A },  /*  87:   W  */
+    { 0x42, 0x42 },  /*  88:   X  */
+    { 0x02, 0x42 },  /*  89:   Y  */
+    { 0x86, 0x86 },  /*  90:   Z  */
+    { 0x00, 0xAC },  /*  91:   [  */
+    { 0x40, 0x40 },  /*  92:   \  */
+    { 0xAC, 0x00 },  /*  93:   ]  */
+    { 0x00, 0x00 },  /*  94:   ^  */
+    { 0x80, 0x04 },  /*  95:   _  */
+    { 0x00, 0x40 },  /*  96:   `  */
+    { 0x00, 0x00 },  /*  97:   a  */
+    { 0x00, 0x00 },  /*  98:   b  */
+    { 0x00, 0x00 },  /*  99:   c  */
+    { 0x00, 0x00 },  /* 100:   d  */
+    { 0x00, 0x00 },  /* 101:   e  */
+    { 0x00, 0x00 },  /* 102:   f  */
+    { 0x00, 0x00 },  /* 103:   g  */
+    { 0x00, 0x00 },  /* 104:   h  */
+    { 0x00, 0x00 },  /* 105:   i  */
+    { 0x00, 0x00 },  /* 106:   j  */
+    { 0x00, 0x00 },  /* 107:   k  */
+    { 0x00, 0x00 },  /* 108:   l  */
+    { 0x00, 0x00 },  /* 109:   m  */
+    { 0x00, 0x00 },  /* 110:   n  */
+    { 0x00, 0x00 },  /* 111:   o  */
+    { 0x00, 0x00 },  /* 112:   p  */
+    { 0x00, 0x00 },  /* 113:   q  */
+    { 0x00, 0x00 },  /* 114:   r  */
+    { 0x00, 0x00 },  /* 115:   s  */
+    { 0x00, 0x00 },  /* 116:   t  */
+    { 0x00, 0x00 },  /* 117:   u  */
+    { 0x00, 0x00 },  /* 118:   v  */
+    { 0x00, 0x00 },  /* 119:   w  */
+    { 0x00, 0x00 },  /* 120:   x  */
+    { 0x00, 0x00 },  /* 121:   y  */
+    { 0x00, 0x00 },  /* 122:   z  */
+    { 0x85, 0x11 },  /* 123:   {  */
+    { 0x01, 0x01 },  /* 124:   |  */
+    { 0x11, 0x85 },  /* 125:   }  */
+    { 0x00, 0x00 },  /* 126:   ~  */
+    { 0x00, 0x00 },  /* 127:      */
+};
+
+/* Internal function */
+static void _shift_char(dsp0401_t *dev, uint8_t c);
+static void _latch(dsp0401_t *dev);
+
+/*---------------------------------------------------------------------------*
+ *                          DSP0401 Core API                                 *
+ *---------------------------------------------------------------------------*/
+
+int dsp0401_init(dsp0401_t *dev, const dsp0401_params_t *params)
+{
+    dev->params = *params;
+
+    /* Initialize clock pin and set it to low */
+    gpio_init(SCK, GPIO_OUT);
+    gpio_clear(SCK);
+
+    /* Initialize data pin */
+    gpio_init(SDI, GPIO_OUT);
+
+    /* Initialize latch pin and set it to low */
+    gpio_init(LAT, GPIO_OUT);
+    gpio_clear(LAT);
+
+    /* Initialize PWM used for LED brightness */
+    pwm_init(PWM_PERIPH, PWM_LEFT, PWM_FREQ, PWM_STEPS);
+    pwm_set(PWM_PERIPH, PWM_CHAN, BRIGHTNESS);
+
+    dsp0401_clear_text(dev);
+
+    return DSP0401_OK;
+}
+
+void dsp0401_display_text(dsp0401_t *dev, char * text)
+{
+    for (size_t i = 0 ; i < MOD_COUNT * 4; ++i) {
+        if (i < strlen(text)) {
```

Comment:
its quite expensive to run `strlen(text)` for every loop count ... better get once and store